backend/auth/role_map_repository.go

```diff
@@ -0,0 +1,279 @@
+	})
+
+	if instance == nil {
+		return nil, errors.New("failed to initialize RoleMapRepository")
+	}
+
+	return instance, nil
+}
+
+func (rmr *RoleMapRepository) HasPermission(rolenames []string, operation *models.Operation) bool {
+	visited := make(map[string]struct{})
+	for _, role := range rolenames {
+		role := rmr.RoleMap[role]
+		if hasPermission(role, rmr.SubroleMap, operation, visited) {
+			return true
+		}
+	}
+	return false
+}
+
+func (rmr *RoleMapRepository) AddFlattenedMap() {
+	rmr.flattenedMap = flattenRoleMap(rmr.RoleMap, rmr.SubroleMap)
+}
+
+
+func flattenRoleMap (roleMap map[string]*models.Role, subroleMap map[string]*models.Role) map[string]map[string]map[string][]models.OperationType{
+	flattenedMap := make(map[string]map[string]map[string][]models.OperationType)
+	for rolename, role := range roleMap {
+		oneRoleMap := flattenRole(role, subroleMap)
+		flattenedMap[rolename] = oneRoleMap
+	}
+	return flattenedMap
+}
+
+func flattenRole(role *models.Role, subroleMap map[string]*models.Role) map[string]map[string][]models.OperationType {
+	namespaces := findUsedOperationsInGraph(
+		role, 
+		subroleMap, 
+		func (op *models.Operation) string { return op.Namespace },
+		make(map[string]struct{}), 
+		make(map[string]struct{}),
+	)
+
+	resources := findUsedOperationsInGraph(
+		role, 
+		subroleMap, 
+		func (op *models.Operation) string { return op.Resource },
+		make(map[string]struct{}), 
+		make(map[string]struct{}),
+	)
+
+	opTypes := models.GetAllOperationTypes()
+
+	result := make(map[string]map[string][]models.OperationType)
+
+	for namespace := range namespaces {
+		result[namespace] = make(map[string][]models.OperationType)
+		for resource := range resources {
+			result[namespace][resource] = make([]models.OperationType, 0, len(opTypes))
+			for _, opType := range opTypes {
+				op := models.Operation{
+					Namespace: namespace,
+					Type: opType,
+					Resource: resource,
+				}
+
+				if hasPermission(role, subroleMap, &op, make(map[string]struct{})) {
+					result[namespace][resource] = append(result[namespace][resource], opType)
+				}
+			}
+		}
+	}
+
+	return result
+}
+
+type getFieldFromOp func(*models.Operation) string
+
+func findUsedOperationsInGraph(
+	role *models.Role,
+	subroleMap map[string]*models.Role, 
+	fieldFunction getFieldFromOp, 
+	values map[string]struct{}, 
+	visited map[string]struct{},
+) map[string]struct{}{
+	for _, op := range role.Deny {
+		value := fieldFunction(&op)
+		values[value] = struct{}{}
+	}
+
+	for _, op := range role.Permit {
+		value := fieldFunction(&op)
+		values[value] = struct{}{}
+	}
+
+	for _, subroleName := range role.Subroles {
+		subrole := subroleMap[subroleName]
+		findUsedOperationsInGraph(subrole, subroleMap, fieldFunction, values, visited)
+	}
+
+	return values
+}
+	
+
+func hasPermission(role *models.Role, subroleMap map[string]*models.Role, operation *models.Operation, visited map[string]struct{}) bool {
```

Comment:
It is intentional as this is helper method for HasPermission. HasPermission is available outside the module while and must be called on the role map instance, so I don't see the way it can be mixed up